contracts/modules/governance/NXMaster.sol

```diff
@@ -47,152 +45,118 @@ contract NXMaster is INXMMaster, Governed {
-      isProxy[_contractName] = true;
-    }
-    allContractVersions[_contractName] = address(uint160(newInternalContract));
-    contractsActive[newInternalContract] = true;
-    LegacyMasterAware up = LegacyMasterAware(allContractVersions[_contractName]);
-    up.changeMasterAddress(address(this));
-    up.changeDependentContractAddress();
-  }
-
-  /**
-   * @dev Anyone can close a claim if oraclize fails to close it.
-   * @param _claimId id of claim to be closed.
-   */
-  function closeClaim(uint _claimId) external {
-
-    require(canCall(_claimId), "Payout retry time not reached.");
-    IClaimsReward cr = IClaimsReward(getLatestAddress("CR"));
-    cr.changeClaimStatus(_claimId);
-  }
-
-  function getOwnerParameters(bytes8 code) external view returns (bytes8 codeVal, address val) {
-    codeVal = code;
-    IQuotationData qd;
-    LegacyPoolData pd;
-    if (code == "MSWALLET") {
-      ITokenData td;
-      td = ITokenData(getLatestAddress("TD"));
-      val = td.walletAddress();
-
-    } else if (code == "MCRNOTA") {
 
-      pd = LegacyPoolData(getLatestAddress("PD"));
-      val = pd.notariseMCR();
-
-    } else if (code == "OWNER") {
+    require(contractAddresses[contractCode] == address(0), "NXMaster: code already in use");
+    require(contractAddress != address(0), "NXMaster: contract address is 0");
 
-      val = owner;
+    contractCodes.push(contractCode);
 
-    } else if (code == "QUOAUTH") {
-
-      qd = IQuotationData(getLatestAddress("QD"));
-      val = qd.authQuoteEngine();
+    address newInternalContract;
+    if (_type == 1) {
 
-    } else if (code == "KYCAUTH") {
-      qd = IQuotationData(getLatestAddress("QD"));
-      val = qd.kycAuthAddress();
+      newInternalContract = contractAddress;
+      isUpgradable[contractCode] = true;
+    } else if (_type == 2) {
 
+      newInternalContract = address(new OwnedUpgradeabilityProxy(contractAddress));
+      isProxy[contractCode] = true;
+    } else {
+      revert("NXMaster: Unsupported contract type");
     }
 
-  }
-
-  /// @dev Add Emergency pause
-  /// @param _pause to set Emergency Pause ON/OFF
-  /// @param _by to set who Start/Stop EP
-  function addEmergencyPause(bool _pause, bytes4 _by) public onlyAuthorizedToGovern {
-    emergencyPaused.push(EmergencyPause(_pause, now, _by));
-    if (_pause == false) {
-      IClaims c1 = IClaims(allContractVersions["CL"]);
-      c1.submitClaimAfterEPOff(); // Process claims submitted while EP was on
-      c1.startAllPendingClaimsVoting(); // Resume voting on all pending claims
-    }
-  }
+    contractAddresses[contractCode] = address(uint160(newInternalContract));
+    contractsActive[newInternalContract] = true;
 
-  ///@dev update time in seconds for which emergency pause is applied.
-  function updatePauseTime(uint _time) public {
-    require(isInternal(msg.sender), "Not internal call.");
-    pauseTime = _time;
+    MasterAware up = MasterAware(contractAddresses[contractCode]);
```

Comment:
Use `newInternalContract` instead of `contractAddresses[contractCode]` to avoid extra storage read.